### PR TITLE
common/buffer: adjust align before calling posix_memalign()

### PR DESCRIPTION
**Pull request by idryomov**
posix_memalign() requires alignment argument to be a multiple of
sizeof(void *).  Since it is an implementation detail of buffer,
it needs to be adjusted there -- buffer consumers have no way of
knowing that passing e.g. align == 4 is incorrect.

One place already does the adjustment, but only for align == 0.
The other just asserts.  Fix both and remove the "power of two"
assertion.  Let posix_memalign() return EINVAL and handle that
by throwing buffer::bad_alloc, as expected by the consumers.

Fixes: https://tracker.ceph.com/issues/50646
Signed-off-by: Ilya Dryomov <idryomov@gmail.com>

### CODE REVIEW

File: src/common/buffer.cc [hunk 1]
```diff
@@ -165,8 +166,10 @@ static ceph::spinlock debug_lock;
     MEMPOOL_CLASS_HELPERS();
 
     raw_posix_aligned(unsigned l, unsigned _align) : raw(l) {
+      // posix_memalign() requires a multiple of sizeof(void *)
+      if (_align < sizeof(void *))
+	_align = sizeof(void *);
```
Comment (rzarzynski) [inline src/common/buffer.cc:171]: nit: how about replacing this conditional with `std::max(align, sizeof(void*))` directly on the initializer list?

File: src/common/buffer.cc [hunk 2]
```diff
@@ -103,8 +103,9 @@ static ceph::spinlock debug_lock;
 	   unsigned align,
 	   int mempool = mempool::mempool_buffer_anon)
     {
-      if (!align)
-	align = sizeof(size_t);
+      // posix_memalign() requires a multiple of sizeof(void *)
+      if (align < sizeof(void *))
+	align = sizeof(void *);
```
Comment (rzarzynski) [inline src/common/buffer.cc:108]: nit: `std::max()`?

**Review by rzarzynski — APPROVED** [review]

File: src/common/buffer.cc (hunk 2 above)
Comment (idryomov) [reply src/common/buffer.cc]: done

**Review by rzarzynski — APPROVED** [review]

File: src/common/buffer.cc (hunk 1 above)
Comment (idryomov) [reply src/common/buffer.cc]: I don't like putting logic in the initializer list.  I did it in the constructor body instead.

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2021-05-05_13:07:31-rados-wip-kefu-testing2-2021-05-05-1748-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-05-07_01:47:01-rados-wip-kefu-testing2-2021-05-05-1748-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-05-08_03:28:06-rados-wip-kefu-testing2-2021-05-07-1516-distro-basic-smithi/

**Pull request merged**